File: sdk/go/watch.go
package holt

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"iter"
	"os/exec"
)

// Watch runs `holt watch --json` as a range-over-func iterator of typed
// lines: a hello, then a sync burst for every lane already alive, ready,
// then live changes for as long as you keep ranging (SPEC.md §14.3 step
// 2). This is the primitive onOpen/onParked/... callback-style APIs are
// built from (SPEC.md §14.2) — see WatchLane for a version scoped to one
// lane's Path.
//
//	for line, err := range c.Watch(ctx) {
//	    if err != nil { ...; break }
//	    if line.Kind == holt.WatchCreated { fmt.Println("new lane:", line.Lane.Name) }
//	}
//
// Unlike the TS/Python SDKs, there is no separate "stop iterating" verb:
// breaking out of the range loop, or canceling ctx, both kill the
// underlying process. Because the iterator body runs synchronously on the
// caller's own goroutine (Go 1.23's iter.Seq2 contract), this needs no
// channel or goroutine bridging to get an async-generator-shaped API —
// the scanner loop below IS the generator.
func (c *Client) Watch(ctx context.Context) iter.Seq2[WatchLine, error] {
	return func(yield func(WatchLine, error) bool) {
		cmd := c.command(ctx, "watch", "--json")
		stdout, err := cmd.StdoutPipe()
		if err != nil {
			yield(WatchLine{}, err)
			return
		}
		var stderrBuf bytes.Buffer
		cmd.Stderr = &stderrBuf
		cmd.Stdin = nil

		if err := cmd.Start(); err != nil {
			yield(WatchLine{}, err)
			return
		}
		waited := false
		defer func() {
			_ = cmd.Process.Kill() // no-op if already exited
			if !waited {
				_ = cmd.Wait()
			}
		}()

		scanner := bufio.NewScanner(stdout)
		// A lane can carry a full branch/path/verdict record; give lines
		// more headroom than bufio's 64KiB default before it errors out.
		scanner.Buffer(make([]byte, 64*1024), 1024*1024)

		for scanner.Scan() {
			line := bytes.TrimSpace(scanner.Bytes())
			if len(line) == 0 {
				continue
			}
			var wl WatchLine
			if err := json.Unmarshal(line, &wl); err != nil {
				if !yield(WatchLine{}, err) {
					return
				}
				continue
			}
			if !yield(wl, nil) {
				return
			}
		}

		if ctx.Err() != nil {
			return // caller canceled ctx — that's why the stream ended, not a failure to report
		}
		if err := scanner.Err(); err != nil {
			yield(WatchLine{}, err)
			return
		}
		waited = true
		if err := cmd.Wait(); err != nil {
			code := ExitUsage
			var exitErr *exec.ExitError
			if asExitError(err, &exitErr) {
				code = ExitCode(exitErr.ExitCode())
			}
			yield(WatchLine{}, &Error{Code: code, Stderr: stderrBuf.String(), Command: []string{c.bin(), "watch", "--json"}})
		}
	}
}

// WatchLane is Watch, filtered to events about one lane (Lane.Path) and
// stripped of the hello/ready framing that names no lane — the shape an
// embedder holding one session per lane usually wants: "tell me when THIS
// lane's state changes." A sync event for the lane still passes through:
// it's how a caller that started watching after the lane went live learns
// it exists at all. Compare full paths, not names: names aren't unique
// across repos, but a checkout path is the registry's own primary key
// (SPEC.md §2.1).
//
// It yields WatchEvent, not WatchLine: hello is filtered out here, so the
// header-only fields can't be populated and shouldn't be in the type. Same
// contract as watchLane in the TS/Python/Swift SDKs.
func (c *Client) WatchLane(ctx context.Context, path string) iter.Seq2[WatchEvent, error] {
	return func(yield func(WatchEvent, error) bool) {
		for line, err := range c.Watch(ctx) {
			if err != nil {
				if !yield(WatchEvent{}, err) {
					return
				}
				continue
			}
			event, ok := line.Event()
			if !ok || event.Kind == WatchReady {
				continue
			}
			if event.Lane != nil && event.Lane.Path == path {
				if !yield(event, nil) {
					return
				}
			}
		}
	}
}

File: sdk/python/src/holt/watch.py
from __future__ import annotations

import asyncio
import json
from typing import AsyncGenerator, Optional

from .exec import RunOptions, merged_env
from .types import WatchEvent, WatchLine, parse_watch_line


async def watch_all(opts: Optional[RunOptions] = None) -> AsyncGenerator[WatchLine, None]:
    """`holt watch --json` as an async iterator of typed lines. One object
    per NDJSON line on stdout, in order: `hello`, a `sync` burst for every
    lane already alive, `ready`, then live changes for as long as the
    process runs (SPEC.md §14.3 step 2).

    The child process is killed when you stop consuming — `break` out of an
    `async for`, or call `.aclose()` on the generator. Async generators are
    not guaranteed to be finalized promptly by garbage collection the way
    CPython's refcounting closes sync generators, so wrap long-lived use in
    `contextlib.aclosing()` if you want the subprocess torn down
    deterministically rather than on the next GC pass:

    ```python
    from contextlib import aclosing

    async with aclosing(holt.watch()) as stream:
        async for line in stream:
            ...
    ```

    There is no other way to stop it short: `watch` has no built-in end
    condition, by design (SPEC.md §14).
    """
    opts = opts or RunOptions()
    bin_ = opts.bin or "holt"
    proc = await asyncio.create_subprocess_exec(
        bin_,
        "watch",
        "--json",
        cwd=opts.cwd,
        env=merged_env(opts.env),
        stdin=asyncio.subprocess.DEVNULL,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    assert proc.stdout is not None
    try:
        while True:
            raw = await proc.stdout.readline()
            if not raw:
                break
            line = raw.decode().strip()
            if not line:
                continue
            yield parse_watch_line(json.loads(line))
    finally:
        if proc.returncode is None:
            proc.kill()
            await proc.wait()


async def watch_lane(path: str, opts: Optional[RunOptions] = None) -> AsyncGenerator[WatchEvent, None]:
    """{watch_all}, filtered to events about one lane (`event.lane.path`)
    and stripped of the `hello`/`ready` framing that names no lane — the
    shape an embedder holding one session per lane usually wants: "tell me
    when THIS lane's state changes."

    A `sync` event for the lane still passes through — it is NOT framing.
    It's how a caller that started watching after the lane went live learns
    the lane exists at all, so a match over `event.kind` needs a `sync` arm.

    Compare full paths, not names: names aren't unique across repos, but a
    checkout path is the registry's own primary key (SPEC.md §2.1).
    """
    async for line in watch_all(opts):
        if isinstance(line, WatchEvent) and line.lane is not None and line.lane.path == path:
            yield line
